Moving release process to Bintray

langsmith opened the pull request:
This pr resolves https://github.com/mapbox/mapbox-java/issues/995 by moving the Java SDK artifact release process to Bintray. The work in this pr is largely based off of [similar work done in `mapbox-gl-native`](https://github.com/mapbox/mapbox-gl-native/pull/13999) and discussion that's happened in https://github.com/mapbox/mapbox-java/issues/995.

langsmith commented:
I believe the only remaining work on this pr is figuring out how to delineate whether a version is a SNAPSHOT or a stable/alpha version as mentioned by @LukasPaczos at https://github.com/mapbox/mapbox-java/issues/995#issuecomment-490910570

langsmith commented:
> figuring out how to delineate whether a version is a SNAPSHOT or a stable/alpha version

https://github.com/mapbox/mapbox-java/pull/1037/commits/9f5db8378e4bf032f9d56e9ff329b7f520f96a66 added logic to `circle.yml` to determine this.

langsmith commented:
@LukasPaczos , I need to request OSS Jfrog access for SNAPSHOTs of each of the 4 Java packages https://bintray.com/message/addPackageToJCenter?pkgPath=%2Fmapbox%2Fmapbox%2Fmapbox-sdk-turf&tab=general 👇 
<img width="986" alt="Screen Shot 2019-05-17 at 4 59 05 PM" src="https://user-images.githubusercontent.com/4394910/57961640-29ff5300-78c5-11e9-8512-05a85bca6758.png">


When you did this for the Maps SDK, do you remember what you checked/entered?

I tried 👇 
<img width="963" alt="Screen Shot 2019-05-17 at 5 01 09 PM" src="https://user-images.githubusercontent.com/4394910/57961673-65018680-78c5-11e9-9b00-d515dbb7c1d3.png">

However, that gives me the following error in the red box. Have you seen this before? What was your work around, if any?

<img width="818" alt="Screen Shot 2019-05-17 at 5 01 56 PM" src="https://user-images.githubusercontent.com/4394910/57961724-7fd3fb00-78c5-11e9-9624-0fcadb6a3aa2.png">

langsmith commented:
Or can I make up some sort of unique ID such as `com.mapbox.mapboxsdk.mapbox-sdk-turf`?

https://www.jfrog.com/confluence/display/RTF/Deploying+Snapshots+to+oss.jfrog.org#DeployingSnapshotstooss.jfrog.org-GettingStartedwithOJO

LukasPaczos commented:
I did those exact steps while setting up Maps SDK. It's weird that you wouldn't be able to publish another artifact under the same Group ID to Artifactory. Could you try pushing them anyway? [The package structure is established on JCenter](http://oss.jfrog.org/oss-snapshot-local/com/mapbox/mapboxsdk/), so maybe it'll pick it up?

If not, the best way would be to contact support - maybe that's not an expected use-case.

langsmith commented:
Hm, ok. Will look into it @LukasPaczos . Contacted support as backup.

LukasPaczos commented:
Any progress here @langsmith? Have you tried pushing the `SNAPSHOT`? Any news from the support?

langsmith commented:
> news from the support

No news, probably because we're not Enterprise customers, so there's no urgency on their part.

> Could you try pushing them anyway
> Have you tried pushing the SNAPSHOT

You mean publishing the test SNAPSHOT as if it were a stable release, correct?

langsmith commented:
>Could you try pushing them anyway
> Have you tried pushing the SNAPSHOT

As a cautious first step I just did a test run of the 4 modules. Then I published `4.9.0-SNAPSHOT` of only Turf rather than all 4 modules: https://bintray.com/mapbox/mapbox/mapbox-sdk-turf/4.9.0-SNAPSHOT

When I clicked on `Click here to get it included.` link 
![Screen Shot 2019-05-24 at 3 05 09 PM](https://user-images.githubusercontent.com/4394910/58358825-68e95780-7e35-11e9-9b33-a3ebabac41ab.png)

I got the following error message

![Screen Shot 2019-05-24 at 3 05 02 PM](https://user-images.githubusercontent.com/4394910/58358834-743c8300-7e35-11e9-94a6-aff8522867af.png)

langsmith commented:
This is the Turf POM file

![Screen Shot 2019-05-24 at 3 07 35 PM](https://user-images.githubusercontent.com/4394910/58358895-ae0d8980-7e35-11e9-9182-e4eaa9af101a.png)

langsmith commented:
Think I found at least one error with the POM file. Looks like the Turf POM file above incorrectly has `services-geojson` for the geojson dependency artifact id. Based on https://search.maven.org/search?q=Mapbox, it should be `mapbox-sdk-geojson`.

langsmith commented:
Ok, I adjusted so that the correct `artifact-id` is added to the POM file.
https://github.com/mapbox/mapbox-java/blob/1cc1392edc1f62ea298a549c23b706f81ef4750b/gradle/gradle-bintray.gradle#L13


Deleted all other Java SDK files from Bintray, built release files on my local machine and then re-published to Bintray. Better Turf POM file is created:

![Screen Shot 2019-05-24 at 3 45 16 PM](https://user-images.githubusercontent.com/4394910/58359971-efecfe80-7e3a-11e9-9b4a-d078b7b075a3.png)

`services` module POM file:

![Screen Shot 2019-05-24 at 3 49 13 PM](https://user-images.githubusercontent.com/4394910/58360068-7e618000-7e3b-11e9-8f95-1500bd2c0f2b.png)

But darn. Despite all of this, still getting the error message.

![Screen Shot 2019-05-24 at 3 50 30 PM](https://user-images.githubusercontent.com/4394910/58360108-b4066900-7e3b-11e9-9db6-c03962d8beb7.png)

langsmith commented:
Do the POM files need a `<packaging>jar</packaging>` line? Doesn't seem to be in https://search.maven.org/artifact/com.mapbox.mapboxsdk/mapbox-sdk-services/4.8.0/jar

langsmith commented:
So weird. 

https://search.maven.org/artifact/com.mapbox.mapboxsdk/mapbox-sdk-services/4.8.0/jar compared with the `services` POM file above (found at https://bintray.com/mapbox/mapbox/download_file?file_path=com%2Fmapbox%2Fmapboxsdk%2Fmapbox-sdk-services%2F4.9.0-SNAPSHOT%2Fmapbox-sdk-services-4.9.0-SNAPSHOT.pom) , reveals no major differences. The name, description, and url tags are in a different location. The bintray POM doesn't have the test scope dependencies either. Not sure whether any of those would lead to the red Bintray error message.

LukasPaczos commented:
> You mean publishing the test SNAPSHOT as if it were a stable release, correct?

No, I meant publishing a `SNAPSHOT` to an actual `SNAPSHOT` repository to test possible JCenter compatibility ([publishing to artifactory job](https://github.com/mapbox/mapbox-java/pull/1037/files#diff-49e0bb33be7766b39a755e0ae6f81b39R73-R85)). The POM does look fine, and I see the "4.9.0-SNAPSHOT" marked as released https://bintray.com/mapbox/mapbox/mapbox-sdk-turf/4.9.0-SNAPSHOT, it also seems to be resolvable by Gradle (well, it's missing `mapbox-sdk-geojson`, but the `mapbox-sdk-turf` looks fine). Did you succeed after all?

langsmith commented:
No news from Bintray support. I don't expect any to arrive.

Tried `make build-release` and then `make publish-snapshot`. Got the following 401 for each of the 4 modules:

<img width="1184" alt="Screen Shot 2019-05-29 at 8 27 25 PM" src="https://user-images.githubusercontent.com/4394910/58606506-86149080-8250-11e9-8565-0c9b1af06d77.png">

I haven't tried accessing the SNAPSHOT via an actual project though.

LukasPaczos commented:
Since we are still blocked on the JCenter front, I'd suggest removing all `SNAPSHOT` changes and focus on the MavenCentral release flow. We can still publish snapshots to Sonatype and revisit the setup when we know more about the JCenter repositories. What do you think @langsmith?

langsmith commented:
I was testing with `4.9.0-SNAPSHOT`. The `-SNAPSHOT` ending isn't allowed in the pom file 🤦‍♂ 

https://central.sonatype.org/pages/requirements.html:
<img width="965" alt="Screen Shot 2019-07-01 at 1 20 59 PM" src="https://user-images.githubusercontent.com/4394910/60465535-2ec65f00-9c06-11e9-8677-4bbcff59dcd2.png">

When I switched to `4.9.0` for a **_test_** run of a new release, the MavenCentral request successfully went through:
<img width="632" alt="Screen Shot 2019-07-01 at 1 19 44 PM" src="https://user-images.githubusercontent.com/4394910/60465573-47367980-9c06-11e9-9ac3-e253fe8a86c5.png">
<img width="826" alt="Screen Shot 2019-07-01 at 1 16 28 PM" src="https://user-images.githubusercontent.com/4394910/60465575-47367980-9c06-11e9-8183-0cc9d0bc7f53.png">

Now we wait for the Bintray team to accept my request so that we have the option to sync the 4 modules' files with MavenCentral 🤞 I won't actually sync because this wasn't a true `4.9.0` release.

langsmith commented:
> Now we wait for the Bintray team to accept my request so that we have the option to sync the 4 modules' files with MavenCentral

They gave us permission and we're now able to sync files with MavenCentral. SNAPSHOTs also look setup in our Bintray account.  This pr is ready for review.

LukasPaczos commented on Makefile:
```diff
@@ -27,18 +27,25 @@ javadoc:
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services-turf:artifactoryPublish ; \
+
 
 publish-local:
```
This won't work and can be removed.

langsmith replied:
Removed

LukasPaczos commented on build.gradle:
```diff
@@ -3,6 +3,7 @@ buildscript {
 
   repositories {
     maven { url 'https://plugins.gradle.org/m2' }
+    maven { url 'https://oss.jfrog.org/artifactory/oss-snapshot-local/' }
```
This is not needed and can be removed.

langsmith replied:
Removed

LukasPaczos commented on circle.yml:
```diff
@@ -58,20 +58,23 @@ jobs:
           command: |
-            if [ "${CIRCLE_BRANCH}" == "master" ]; then
-            make publish ;
+            if [[ $CIRCLE_BRANCH == master ]] || [[ $CIRCLE_BRANCH == release-* ]] || [[ $CIRCLE_TAG == android-v* ]]; then
```
Tag convention for this repo is `v[number]`, others can be removed (besides `master`).

langsmith replied:
Fixed

LukasPaczos commented on services-core/gradle.properties:
```diff
@@ -1,5 +1,5 @@
 VERSION_NAME=4.9.0-SNAPSHOT
 POM_ARTIFACT_ID=mapbox-sdk-core
 POM_NAME=Mapbox Services SDK
```
This seems unused.

langsmith replied:
Removed `POM_NAME` from each `gradle.properties` file

langsmith commented:
> Also, the readme should be updated with the new snapshot repository URL.

Updated

langsmith commented:
>The release process is missing the step that changes the version name and removes the SNAPSHOT reference in order to support a full tag release. Check out this step. This should be applied for each artifact.

Added the step to `circle.yml` and wrote it the way that I did because I deleted `VERSION_NAME` from each nested `gradle.properties` file. The `VERSION_NAME` is the same for all 4 modules/artifacts and it's already declared in the top level `gradle.properties` file.

langsmith commented:
Ok @LukasPaczos , pushed a commit that addressed your latest review.

LukasPaczos commented on circle.yml:
```diff
@@ -58,20 +58,29 @@ jobs:
           command: |
-            if [ "${CIRCLE_BRANCH}" == "master" ]; then
-            make publish ;
+            if [[ $CIRCLE_BRANCH == master ]] || [[ $CIRCLE_TAG == android-v* ]]; then
```
`[[ $CIRCLE_TAG == android-v* ]]` should be `[[ $CIRCLE_TAG == v* ]]`

langsmith replied:
Fixed

LukasPaczos commented on build.gradle:
```diff
@@ -97,12 +102,14 @@ subprojects { subproject ->
 
   if (RELEASE_MODULES.contains(subproject.name)) {
-    subproject.apply from: "${rootDir}/gradle/mvn-push.gradle"
+    subproject.apply plugin: 'com.jfrog.bintray'
```
Already applied in the `gradle-bintray.gradle`.

langsmith replied:
Removed

LukasPaczos commented on build.gradle:
```diff
@@ -97,12 +102,14 @@ subprojects { subproject ->
   if (RELEASE_MODULES.contains(subproject.name)) {
-    subproject.apply from: "${rootDir}/gradle/mvn-push.gradle"
+    subproject.apply plugin: 'com.jfrog.bintray'
+    subproject.apply plugin: "com.jfrog.artifactory"
```
Same as above.

langsmith replied:
Removed

LukasPaczos commented on build.gradle:
```diff
@@ -97,12 +102,14 @@ subprojects { subproject ->
-    subproject.apply from: "${rootDir}/gradle/mvn-push.gradle"
+    subproject.apply plugin: 'com.jfrog.bintray'
+    subproject.apply plugin: "com.jfrog.artifactory"
+    subproject.apply from: "${rootDir}/gradle/artifact-settings.gradle"
```
Same as above.

langsmith replied:
Removed

LukasPaczos commented on Makefile:
```diff
@@ -27,18 +27,17 @@ javadoc:
-    export IS_LOCAL_DEVELOPMENT=true; ./gradlew :services:uploadArchives ; \
-    export IS_LOCAL_DEVELOPMENT=true; ./gradlew :services-turf:uploadArchives ; \
+publish-to-bintray:
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services-core:bintrayUpload ; \
```
`IS_LOCAL_DEVELOPMENT` is not used.

langsmith replied:
Removed

tobrun commented:
@langsmith tried creating a snapshot but hitting:

Task :artifactoryDeploy FAILED
Deploying artifact: http://oss.jfrog.org/oss-snapshot-local/com/mapbox/mapboxsdk/mapbox-sdk-core/4.9.2-SNAPSHOT/mapbox-sdk-core-4.9.2-SNAPSHOT.jar

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':artifactoryDeploy'.
> java.io.IOException: Failed to deploy file. Status code: 403 Response message: Artifactory returned the following errors:
 User mapbox-bot is not permitted to deploy 'com/mapbox/mapboxsdk/mapbox-sdk-core/4.9.2-SNAPSHOT/mapbox-sdk-core-4.9.2-SNAPSHOT.jar' into 'oss-snapshot-local:com/mapbox/mapboxsdk/mapbox-sdk-core/4.9.2-SNAPSHOT/mapbox-sdk-core-4.9.2-SNAPSHOT.jar'. Status code: 403

langsmith commented:
Java SDK SNAPSHOTs should be working and available now @tobrun https://oss.jfrog.org/artifactory/oss-snapshot-local/com/mapbox/mapboxsdk/

<img width="942" alt="Screen Shot 2019-07-09 at 5 55 58 PM" src="https://user-images.githubusercontent.com/4394910/60932419-d250e880-a272-11e9-919c-bc6501cfee3f.png">


<img width="1059" alt="Screen Shot 2019-07-09 at 5 57 41 PM" src="https://user-images.githubusercontent.com/4394910/60932483-1217d000-a273-11e9-9401-6b241cff99fb.png">
